curvefs/src/client/fuse_client.cpp

```diff
@@ -946,68 +948,154 @@ CURVEFS_ERROR FuseClient::CalOneLayerSumInfo(Inode *inode) {
-            if (ret == CURVEFS_ERROR::OK) {
-                for (const auto &it : attrs) {
+        return true;
+    }
+}
+
+void FuseClient::ConcurrentGetInodeAttr(
+    std::stack<uint64_t> *iStack,
+    std::mutex *stackMutex,
+    std::unordered_map<uint64_t, uint64_t> *hardLinkMap,
+    std::mutex *mapMutex,
+    bool dirOnly,
+    uint64_t *rfiles,
+    uint64_t *rsubdirs,
+    uint64_t *rentries,
+    uint64_t *rfbytes,
+    std::mutex *valueMutex,
+    std::atomic<uint32_t> *inflightNum,
+    std::atomic<bool> *ret) {
+    while (1) {
+        std::list<Dentry> dentryList;
+        std::set<uint64_t> inodeIds;
+        std::list<InodeAttr> attrs;
+        auto tret = ConcurrentListDentry(&dentryList, iStack, stackMutex,
+                                        dirOnly, inflightNum, ret);
+        if (!tret) {
+            break;
+        } else {
+            {
+                std::lock_guard<std::mutex> guard(*stackMutex);
+                for (const auto &it : dentryList) {
                     if (it.type() == FsFileType::TYPE_DIRECTORY) {
-                        rsubdirs++;
-                    } else {
-                        rfiles++;
+                        iStack->emplace(it.inodeid());
                     }
-                    rentries++;
-                    rfbytes += it.length();
-                    // record hardlink
-                    if (it.type() != FsFileType::TYPE_DIRECTORY &&
-                        it.nlink() > 1) {
-                        if (hardLinkMap.count(it.inodeid())) {
-                            hardLinkMap[it.inodeid()] += it.length();
+                    inodeIds.emplace(it.inodeid());
+                }
+            }
+            inflightNum->fetch_sub(1, std::memory_order_seq_cst);
+            if (!inodeIds.empty()) {
+                auto tret = inodeManager_->BatchGetInodeAttr(&inodeIds, &attrs);
+                if (tret == CURVEFS_ERROR::OK) {
+                    std::lock_guard<std::mutex> guard(*valueMutex);
+                    for (const auto &it : attrs) {
+                        if (it.type() == FsFileType::TYPE_DIRECTORY) {
+                            (*rsubdirs)++;
                         } else {
-                            hardLinkMap.emplace(it.inodeid(), 0);
+                            (*rfiles)++;
+                        }
+                        (*rentries)++;
+                        (*rfbytes) += it.length();
+                        // record hardlink
+                        if (it.type() != FsFileType::TYPE_DIRECTORY &&
+                            it.nlink() > 1) {
+                            std::lock_guard<std::mutex> guard(*mapMutex);
+                            auto iter = hardLinkMap->find(it.inodeid());
+                            if (iter != hardLinkMap->end()) {
+                                iter->second += it.length();
+                            } else {
+                                hardLinkMap->emplace(it.inodeid(), 0);
```

Comment:
why insert a key-value pair and the value is 0 ?

should it be `it.length()` ?

if so, you can simplify these to `hardLinkMap[it.inodeid()] += it.length()`

Reply:
The logic is: if (it.type() != FsFileType::TYPE_DIRECTORY && it.nlink() > 1) means this file is a hardlink, but if others hardlinks of the file not exists in the dir which summary, the size is correct.
 The hardLinkMap <inodeId, the size need to minus>, so the first time is 0.